src/org/traccar/protocol/Arnavi4ProtocolDecoder.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2017 Ivan Muratov (binakot@gmail.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.jboss.netty.channel.Channel;
+import org.traccar.BaseProtocolDecoder;
+import org.traccar.DeviceSession;
+import org.traccar.helper.Checksum;
+import org.traccar.model.Position;
+
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.traccar.protocol.Arnavi4FrameDecoder.HEADER_START_SIGN;
+import static org.traccar.protocol.Arnavi4FrameDecoder.HEADER_VERSION_1;
+import static org.traccar.protocol.Arnavi4FrameDecoder.HEADER_VERSION_2;
+
+public class Arnavi4ProtocolDecoder extends BaseProtocolDecoder {
+
+    private static final byte RECORD_PING = 0x00;
+    private static final byte RECORD_DATA = 0x01;
+    private static final byte RECORD_TEXT = 0x03;
+    private static final byte RECORD_FILE = 0x04;
+    private static final byte RECORD_BINARY = 0x06;
+
+    private static final byte TAG_LATITUDE = 3;
+    private static final byte TAG_LONGITUDE = 4;
+    private static final byte TAG_COORD_PARAMS = 5;
+
+    public Arnavi4ProtocolDecoder(Arnavi4Protocol protocol) {
+        super(protocol);
+    }
+
+    private void sendHeaderResponse(Channel channel, byte version) {
```

Comment:
The idea was to unify all responses, not just extract pieces of code into separate methods. Please avoid using array. Check any protocol encoder for example on how we usually construct a message.